src/secure-proxy-handler.ts

```diff
@@ -109,6 +112,14 @@ export class SecureProxyHandler implements ProxyHandler<SecureProxyTarget> {
         setPrototypeOf(shadowTarget, env.getSecureValue(rawProto));
         // defining own descriptors
         copySecureOwnDescriptors(env, shadowTarget, target);
+        // preserving the semantics of the object
+        if (isFrozen(target)) {
+            freeze(shadowTarget);
+        } else if (isSealed(target)) {
+            seal(shadowTarget);
+        } else if (!isExtensible(target)) {
+            preventExtensions(shadowTarget);
+        }
```

Comment:
Nice logic fall of most restricted to least. ☝️ 👍

Reply:
yeah, trying to avoid the perf hit here, it is still not ideal, maybe we can do this lazily somehow but the proxy's traps do not allow detection of freezing and sealing, only extensibility.